In [19]:
import duckdb
import pandas as pd

con = duckdb.connect("../data/shelterflow.duckdb")
bronze_intakes = con.execute("SELECT * FROM bronze_intakes").df()
bronze_outcomes = con.execute("SELECT * FROM bronze_outcomes").df()
silver_intakes = con.execute("SELECT * FROM silver_intakes").df()
silver_outcomes = con.execute("SELECT * FROM silver_outcomes").df()

pd.set_option("display.max_rows", None)

# `silver_intakes` Validation

In [20]:
silver_intakes.info()

<class 'pandas.DataFrame'>
RangeIndex: 80159 entries, 0 to 80158
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   animal_id           80159 non-null  str           
 1   animal_type         80159 non-null  str           
 2   intake_type         80159 non-null  str           
 3   intake_condition    80159 non-null  str           
 4   sex_upon_intake     80158 non-null  str           
 5   age_upon_intake     80159 non-null  str           
 6   found_location      80159 non-null  str           
 7   color               80159 non-null  str           
 8   name                55585 non-null  str           
 9   breed               80159 non-null  str           
 10  breed_standardized  80159 non-null  str           
 11  age_in_days         80159 non-null  int32         
 12  age_group           75379 non-null  str           
 13  intake_date         80159 non-null  datetime64[us]
dtypes

## Comparing `silver_intakes` against `bronze_intakes`

In [21]:
silver_intakes.head(30)

,animal_id,animal_type,intake_type,intake_condition,sex_upon_intake,age_upon_intake,found_location,color,name,breed,breed_standardized,age_in_days,age_group,intake_date
0,A731435,Cat,Owner Surrender,Normal,Neutered Male,5 months,Austin (TX),Cream Tabby,*Casey,Domestic Shorthair Mix,Domestic Shorthair Mix,150,Kitten,2016-08-08
1,A758038,Dog,Stray,Normal,Intact Female,1 year,7100 Colony Park Drive in Austin (TX),Black/White,NaN,Siberian Husky Mix,Siberian Husky Mix,365,Young,2017-09-10
2,A746979,Dog,Owner Surrender,Normal,Intact Male,1 year,7601 Daffan Ln in Austin (TX),Brown Brindle,Teddy,Chihuahua Shorthair Mix,Chihuahua Shorthair Mix,365,Young,2017-04-12
3,A705868,Dog,Stray,Normal,Intact Male,5 months,22009 Hogeye Rd in Austin (TX),Red,Izzy,Golden Retriever Mix,Golden Retriever Mix,150,Puppy,2015-06-22
4,A714106,Dog,Stray,Normal,Intact Male,1 year,183/Mlk in Travis (TX),White/Brown,*Nexus,Pit Bull Mix,Pit Bull Mix,365,Young,2015-10-18
5,A717905,Dog,Stray,Normal,Intact Male,1 month,11717 Schriber Rd in Travis (TX),Black,NaN,Chesa Bay Retr Mix,Chesa Bay Retr Mix,30,Puppy,2015-12-17
6,A675336,Dog,Stray,Normal,Intact Female,3 years,Iwanna Dr in Del Valle (TX),Silver,Runt Runt,Standard Schnauzer Mix,Standard Schnauzer Mix,1095,Adult,2014-03-26
7,A726125,Dog,Owner Surrender,Normal,Spayed Female,1 year,Austin (TX),Brown,Carley,Pit Bull Mix,Pit Bull Mix,365,Young,2017-04-07
8,A739497,Dog,Stray,Normal,Intact Female,2 years,1515 Wickersham Ln in Austin (TX),Brown,*Millie,Australian Cattle Dog/Beagle,Australian Cattle Dog Mix,730,Young,2016-12-03
9,A734336,Other,Wildlife,Sick,Unknown,1 year,245 W 3Rd St in Austin (TX),Brown,NaN,Rat,Rat,365,NaN,2016-09-05


In [22]:
bronze_intakes.head(30)

,age_upon_intake,animal_id,animal_type,breed,color,datetime,datetime2,found_location,intake_condition,intake_type,name,sex_upon_intake
0,8 years,A706918,Dog,English Springer Spaniel,White/Liver,2015-07-05T12:59:00.000,2015-07-05T12:59:00.000,9409 Bluegrass Dr in Austin (TX),Normal,Stray,Belle,Spayed Female
1,11 months,A724273,Dog,Basenji Mix,Sable/White,2016-04-14T18:43:00.000,2016-04-14T18:43:00.000,2818 Palomino Trail in Austin (TX),Normal,Stray,Runster,Intact Male
2,4 weeks,A665644,Cat,Domestic Shorthair Mix,Calico,2013-10-21T07:59:00.000,2013-10-21T07:59:00.000,Austin (TX),Sick,Stray,NaN,Intact Female
3,4 years,A682524,Dog,Doberman Pinsch/Australian Cattle Dog,Tan/Gray,2014-06-29T10:38:00.000,2014-06-29T10:38:00.000,800 Grove Blvd in Austin (TX),Normal,Stray,Rio,Neutered Male
4,2 years,A743852,Dog,Labrador Retriever Mix,Chocolate,2017-02-18T12:46:00.000,2017-02-18T12:46:00.000,Austin (TX),Normal,Owner Surrender,Odin,Neutered Male
5,2 years,A708452,Dog,Labrador Retriever Mix,Black/White,2015-07-30T14:37:00.000,2015-07-30T14:37:00.000,Austin (TX),Normal,Public Assist,Mumble,Intact Male
6,5 months,A731435,Cat,Domestic Shorthair Mix,Cream Tabby,2016-08-08T17:52:00.000,2016-08-08T17:52:00.000,Austin (TX),Normal,Owner Surrender,*Casey,Neutered Male
7,2 years,A760053,Dog,Chihuahua Shorthair,White/Tan,2017-10-11T15:46:00.000,2017-10-11T15:46:00.000,8800 South First Street in Austin (TX),Normal,Stray,NaN,Intact Male
8,5 months,A707375,Dog,Pit Bull,Brown/White,2015-07-11T18:19:00.000,2015-07-11T18:19:00.000,Galilee Court And Damita Jo Dr in Manor (TX),Normal,Stray,*Candy Cane,Intact Female
9,2 years,A696408,Dog,Chihuahua Shorthair,Tricolor,2015-02-04T12:58:00.000,2015-02-04T12:58:00.000,9705 Thaxton in Austin (TX),Normal,Stray,*Pearl,Intact Female


## Verifying `age_group` populates sensibly

In [23]:
silver_intakes[["age_upon_intake", "age_group"]].value_counts().reset_index(name="count")

,age_upon_intake,age_group,count
0,1 year,Young,12433
1,2 years,Young,10762
2,3 years,Adult,4983
3,1 month,Kitten,4962
4,4 years,Adult,3025
5,5 years,Adult,2720
6,1 month,Puppy,2468
7,4 weeks,Kitten,2210
8,3 weeks,Kitten,2128
9,2 months,Kitten,2066


In [24]:
silver_intakes[(silver_intakes["animal_type"].isin(["Dog", "Cat"])) & (silver_intakes["age_group"].isnull())].shape[0]

0

## Verifying deduplication

In [25]:
silver_intakes.duplicated().sum()

0

# `silver_outcomes` Validation

In [26]:
silver_outcomes.info()

<class 'pandas.DataFrame'>
RangeIndex: 80658 entries, 0 to 80657
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   animal_id           80658 non-null  str           
 1   animal_type         80658 non-null  str           
 2   outcome_type        80644 non-null  str           
 3   outcome_subtype     36886 non-null  str           
 4   sex_upon_outcome    80656 non-null  str           
 5   age_upon_outcome    80650 non-null  str           
 6   name                56101 non-null  str           
 7   color               80658 non-null  str           
 8   breed               80658 non-null  str           
 9   breed_standardized  80658 non-null  str           
 10  age_in_days         80650 non-null  Int32         
 11  age_group           75856 non-null  str           
 12  outcome_date        80658 non-null  datetime64[us]
 13  date_of_birth       80658 non-null  datetime64[us]
dtypes

In [27]:
silver_outcomes.head(30)

,animal_id,animal_type,outcome_type,outcome_subtype,sex_upon_outcome,age_upon_outcome,name,color,breed,breed_standardized,age_in_days,age_group,outcome_date,date_of_birth
0,A661023,Dog,Adoption,Foster,Neutered Male,3 years,Lucky,White,Miniature Poodle Mix,Miniature Poodle Mix,1095,Adult,2013-10-15,2010-08-15
1,A677147,Dog,Return to Owner,NaN,Spayed Female,9 years,Joy,White/Tan,Beagle Mix,Beagle Mix,3285,Senior,2014-04-20,2005-04-20
2,A682239,Cat,Adoption,Foster,Spayed Female,2 months,*Rust,Tortie,Domestic Shorthair Mix,Domestic Shorthair Mix,60,Kitten,2014-08-03,2014-05-25
3,A682584,Dog,Adoption,NaN,Neutered Male,1 year,NaN,Tan/White,Chihuahua Shorthair Mix,Chihuahua Shorthair Mix,365,Young,2014-07-13,2013-06-30
4,A669279,Dog,Adoption,NaN,Neutered Male,10 months,Sagan,Tricolor,Anatol Shepherd Mix,Anatol Shepherd Mix,300,Puppy,2013-12-24,2013-02-17
5,A665982,Dog,Return to Owner,NaN,Spayed Female,5 years,Minnie,White/Tan,Chihuahua Shorthair/Cardigan Welsh Corgi,Chihuahua Shorthair Mix,1825,Adult,2013-10-26,2008-10-25
6,A675669,Dog,Adoption,NaN,Spayed Female,1 year,NaN,Blue Merle/Tan,Queensland Heeler Mix,Australian Cattle Dog Mix,365,Young,2014-04-06,2013-03-31
7,A679717,Cat,Transfer,Partner,Intact Female,3 weeks,NaN,Torbie,Domestic Shorthair Mix,Domestic Shorthair Mix,21,Kitten,2014-05-25,2014-05-03
8,A682002,Cat,Transfer,Partner,Intact Female,2 days,NaN,Calico,Domestic Shorthair Mix,Domestic Shorthair Mix,2,Kitten,2014-06-22,2014-06-20
9,A669791,Cat,Transfer,SCRP,Intact Female,1 year,NaN,Tortie,Domestic Shorthair Mix,Domestic Shorthair Mix,365,Young,2013-12-28,2012-12-27


In [28]:
bronze_outcomes.head(30)

,age_upon_outcome,animal_id,animal_type,breed,color,date_of_birth,datetime,monthyear,name,outcome_subtype,outcome_type,sex_upon_outcome
0,2 weeks,A684346,Cat,Domestic Shorthair Mix,Orange Tabby,2014-07-07T00:00:00,2014-07-22T16:04:00,2014-07-22T16:04:00,NaN,Partner,Transfer,Intact Male
1,1 year,A666430,Dog,Beagle Mix,White/Brown,2012-11-06T00:00:00,2013-11-07T11:47:00,2013-11-07T11:47:00,Lucy,Partner,Transfer,Spayed Female
2,1 year,A675708,Dog,Pit Bull,Blue/White,2013-03-31T00:00:00,2014-06-03T14:20:00,2014-06-03T14:20:00,*Johnny,NaN,Adoption,Neutered Male
3,9 years,A680386,Dog,Miniature Schnauzer Mix,White,2005-06-02T00:00:00,2014-06-15T15:50:00,2014-06-15T15:50:00,Monday,Partner,Transfer,Neutered Male
4,5 months,A683115,Other,Bat Mix,Brown,2014-01-07T00:00:00,2014-07-07T14:04:00,2014-07-07T14:04:00,NaN,Rabies Risk,Euthanasia,Unknown
5,4 months,A664462,Dog,Leonberger Mix,Brown/White,2013-06-03T00:00:00,2013-10-07T13:06:00,2013-10-07T13:06:00,*Edgar,Partner,Transfer,Intact Male
6,1 month,A685067,Cat,Domestic Shorthair Mix,Blue Tabby/White,2014-06-16T00:00:00,2014-08-14T18:45:00,2014-08-14T18:45:00,Lucy,NaN,Adoption,Intact Female
7,3 months,A678580,Cat,Domestic Shorthair Mix,White/Black,2014-03-26T00:00:00,2014-06-29T17:45:00,2014-06-29T17:45:00,*Frida,Offsite,Adoption,Spayed Female
8,1 year,A675405,Cat,Domestic Medium Hair Mix,Black/White,2013-03-27T00:00:00,2014-03-28T14:55:00,2014-03-28T14:55:00,Stella Luna,NaN,Return to Owner,Spayed Female
9,2 years,A673652,Dog,Papillon/Border Collie,Black/White,2012-02-28T00:00:00,2014-03-28T14:39:00,2014-03-28T14:39:00,Fancy,Partner,Transfer,Neutered Male


In [29]:
## Verifying `age_group` populates sensibly

In [30]:
silver_outcomes[["age_upon_outcome", "age_group"]].value_counts().reset_index(name="count")

,age_upon_outcome,age_group,count
0,1 year,Young,12664
1,2 years,Young,10824
2,2 months,Kitten,5977
3,3 years,Adult,5060
4,2 months,Puppy,3352
5,4 years,Adult,3017
6,5 years,Adult,2753
7,3 months,Kitten,2449
8,1 month,Kitten,2028
9,6 years,Adult,1870


In [31]:
silver_outcomes[(silver_outcomes["animal_type"].isin(["Dog", "Cat"])) & (silver_outcomes["age_group"].isnull())].shape[0]

6

## Verifying deduplication

In [32]:
silver_outcomes.duplicated().sum()

0

## Investigating Intake/Outcome Count Mismatch

In [33]:
# Count intakes per animal
intake_counts = silver_intakes.groupby("animal_id").size().rename("n_intakes")

# Count outcomes per animal
outcome_counts = silver_outcomes.groupby("animal_id").size().rename("n_outcomes")

# Combine into one table
counts = pd.concat([intake_counts, outcome_counts], axis=1).fillna(0)

# How many animals have mismatched counts?
mismatched = counts[counts["n_intakes"] != counts["n_outcomes"]]
print(f"Animals with mismatched intake/outcome counts: {len(mismatched)}")
print(f"Total animals: {len(counts)}")
print(f"Percentage mismatched: {len(mismatched) / len(counts) * 100:.1f}%")

Animals with mismatched intake/outcome counts: 1543
Total animals: 73269
Percentage mismatched: 2.1%


In [34]:
# See the breakdown of how counts differ
counts["difference"] = counts["n_intakes"] - counts["n_outcomes"]
print(counts["difference"].value_counts().sort_index())

difference
-1.0     1021
 0.0    71726
 1.0      522
Name: count, dtype: int64


In [35]:
clean = counts[(counts["n_intakes"] == 1) & (counts["n_outcomes"] == 1)]
print(f"Animals with exactly 1 intake and 1 outcome: {len(clean)} ({len(clean)/len(counts)*100:.1f}%)")

Animals with exactly 1 intake and 1 outcome: 65743 (89.7%)


71,726 animals (98%) have matching intake and outcome counts and pair directly by chronological sequence. 522 animals have one unmatched intake (likely still in shelter at the dataset's end), and 1,021 have one unmatched outcome (likely intakes predating the dataset's 2013 start). All mismatches are off-by-one; unmatched events are excluded from stay-level metrics via an inner join, an assumption appropriate for aggregate analysis.
